src/internal/packager/images/pull.go

```diff
@@ -27,329 +31,314 @@ import (
+			}
+			return saved, err
 		}
+		saved[info] = img
+	}
+	return saved, nil
+}
 
-		// Attempt to connect to the local docker daemon.
-		ctx := context.TODO()
-		cli, err := client.NewClientWithOpts(client.FromEnv)
-		if err != nil {
-			return nil, false, fmt.Errorf("docker not available: %w", err)
-		}
-		cli.NegotiateAPIVersion(ctx)
+// SaveConcurrent saves images in a concurrent, bounded manner.
+func SaveConcurrent(ctx context.Context, cl clayout.Path, m map[transform.Image]v1.Image) (map[transform.Image]v1.Image, error) {
+	saved := map[transform.Image]v1.Image{}
 
-		// Inspect the image to get the size.
-		rawImg, _, err := cli.ImageInspectWithRaw(ctx, src)
-		if err != nil {
-			return nil, false, fmt.Errorf("failed to inspect image via docker: %w", err)
-		}
+	var mu sync.Mutex
 
-		// Warn the user if the image is large.
-		if rawImg.Size > 750*1000*1000 {
-			message.Warnf("%s is %s and may take a very long time to load via docker. "+
-				"See https://docs.zarf.dev/docs/faq for suggestions on how to improve large local image loading operations.",
-				src, utils.ByteFormat(float64(rawImg.Size), 2))
-		}
+	eg, ectx := errgroup.WithContext(ctx)
+	eg.SetLimit(10)
 
-		// Use unbuffered opener to avoid OOM Kill issues https://github.com/defenseunicorns/zarf/issues/1214.
-		// This will also take for ever to load large images.
-		if img, err = daemon.Image(reference, daemon.WithUnbufferedOpener()); err != nil {
-			return nil, false, fmt.Errorf("failed to load image from docker daemon: %w", err)
-		}
-	} else {
-		refInfo, err := transform.ParseImageRef(src)
-		if err != nil {
-			return nil, false, err
-		}
-		// Check if we have an image index or manifest list and if so error out
-		if refInfo.Digest != "" && (desc.MediaType == ctypes.OCIImageIndex || desc.MediaType == ctypes.DockerManifestList) {
-			var idx v1.IndexManifest
-			if err := json.Unmarshal(desc.Manifest, &idx); err != nil {
-				return nil, false, fmt.Errorf("%w: %w", lang.ErrUnsupportedImageType, err)
-			}
-			imageOptions := "please select one of the images below based on your platform to use instead"
-			imageBaseName := refInfo.Name
-			if refInfo.Tag != "" {
-				imageBaseName = fmt.Sprintf("%s:%s", imageBaseName, refInfo.Tag)
-			}
-			for _, manifest := range idx.Manifests {
-				imageOptions = fmt.Sprintf("%s\n %s@%s for platform %s", imageOptions, imageBaseName, manifest.Digest, manifest.Platform)
+	for info, img := range m {
+		info, img := info, img
+		eg.Go(func() error {
+			desc, err := partial.Descriptor(img)
+			if err != nil {
+				return err
 			}
-			return nil, false, fmt.Errorf("%w: %s", lang.ErrUnsupportedImageType, imageOptions)
-		}
 
-		// Manifest was found, so use crane to pull the image.
-		if img, err = crane.Pull(src, config.GetCraneOptions(i.Insecure, i.Architectures...)...); err != nil {
-			return nil, false, fmt.Errorf("failed to pull image: %w", err)
-		}
-		cacheImage = true
-	}
+			if err := cl.WriteImage(img); err != nil {
```

Comment:
Are we able to use `AppendImage` here like we do in `SaveSequential`? As opposed to using `WriteImage` + `AppendDescriptor`

Reply:
it was being mean to me when run concurrently, so I swapped to using a mutex on the `AppendDescriptor` because I believe that function is not go-routine safe